### PR TITLE
lso pkg: localvolumeset enhancements, unit test, new methods added

### PR DESCRIPTION
**Pull request by elenagerman**
localvolumeset enhancements, unit test, new methods added:

- WithNodeSelector
- WithStorageClassName
- WithVolumeMode
- WithFSType
- WithMaxDeviceCount
- WithDeviceInclusionSpec
-

### CODE REVIEW

**Comment by elenagerman** [comment]
@trewest @kononovn @achuzhoy @cdvultur PTAL

File: pkg/lso/localvolumeset.go [hunk 1]
```diff
@@ -192,9 +212,170 @@ func (builder *LocalVolumeSetBuilder) Update() (*LocalVolumeSetBuilder, error) {
...
+	builder.Object = builder.Definition
+
 	return builder, err
 }
 
+// WithTolerations sets the localVolumeSet operator's generation.
+func (builder *LocalVolumeSetBuilder) WithTolerations(
+	tolerations []corev1.Toleration) *LocalVolumeSetBuilder {
+	glog.V(100).Infof(
+		"Adding tolerations to localVolumeSet %s in namespace %s; tolerations %v",
```
Comment (achuzhoy) [inline pkg/lso/localvolumeset.go:231]:
```suggestion
		"Adding tolerations %v to localVolumeSet %s in namespace %s",
```

File: pkg/lso/localvolumeset.go [hunk 2]
```diff
@@ -192,9 +212,170 @@ func (builder *LocalVolumeSetBuilder) Update() (*LocalVolumeSetBuilder, error) {
...
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	builder.Definition.Spec.NodeSelector = &nodeSelector
+
+	return builder
+}
+
+// WithStorageClassName sets the localVolumeSet operator's storageClassName.
```
Comment (achuzhoy) [inline pkg/lso/localvolumeset.go:267]:
```suggestion
// WithStorageClassName sets the localVolumeSet's storageClassName.
```

File: pkg/lso/localvolumeset.go [hunk 3]
```diff
@@ -192,9 +212,170 @@ func (builder *LocalVolumeSetBuilder) Update() (*LocalVolumeSetBuilder, error) {
...
+func (builder *LocalVolumeSetBuilder) WithStorageClassName(
+	storageClassName string) *LocalVolumeSetBuilder {
+	glog.V(100).Infof(
+		"Adding storageClassName to localVolumeSet %s in namespace %s; storageClassName %v",
```
Comment (achuzhoy) [inline pkg/lso/localvolumeset.go:271]:
```suggestion
		"Adding storageClassName %v to localVolumeSet %s in namespace %s",
```

File: pkg/lso/localvolumeset.go [hunk 4]
```diff
@@ -192,9 +212,170 @@ func (builder *LocalVolumeSetBuilder) Update() (*LocalVolumeSetBuilder, error) {
...
+
+		return builder
+	}
+
+	builder.Definition.Spec.StorageClassName = storageClassName
+
+	return builder
+}
+
+// WithVolumeMode sets the localVolumeSet operator's volumeMode.
```
Comment (achuzhoy) [inline pkg/lso/localvolumeset.go:291]:
```suggestion
// WithVolumeMode sets the localVolumeSet's volumeMode.
```

File: pkg/lso/localvolumeset.go [hunk 5]
```diff
@@ -192,9 +212,170 @@ func (builder *LocalVolumeSetBuilder) Update() (*LocalVolumeSetBuilder, error) {
...
+func (builder *LocalVolumeSetBuilder) WithVolumeMode(
+	volumeMode lsov1.PersistentVolumeMode) *LocalVolumeSetBuilder {
+	glog.V(100).Infof(
+		"Adding volumeMode to localVolumeSet %s in namespace %s; volumeMode %v",
```
Comment (achuzhoy) [inline pkg/lso/localvolumeset.go:295]:
```suggestion
		"Adding volumeMode %v to localVolumeSet %s in namespace %s",
```

Comment (achuzhoy) [reply pkg/lso/localvolumeset.go:295]: Same idea for the rest

**Comment by elenagerman** [comment]
@trewest @kononovn @achuzhoy @cdvultur PTAL

File: pkg/lso/localvolumeset.go [hunk 6]
```diff
@@ -192,9 +212,170 @@ func (builder *LocalVolumeSetBuilder) Update() (*LocalVolumeSetBuilder, error) {
...
+
+		return nil, err
+	}
+
+	builder.Object = builder.Definition
+
 	return builder, err
 }
 
+// WithTolerations sets the localVolumeSet's generation.
```
Comment (achuzhoy) [inline pkg/lso/localvolumeset.go:227]:
```suggestion
// WithTolerations sets the localVolumeSet's toleration.
```

File: pkg/lso/localvolumeset.go [hunk 7]
```diff
@@ -192,9 +212,170 @@ func (builder *LocalVolumeSetBuilder) Update() (*LocalVolumeSetBuilder, error) {
...
+	glog.V(100).Infof(
+		"Adding tolerations %v to localVolumeSet %s in namespace %s",
+		builder.Definition.Name, builder.Definition.Namespace, tolerations)
+
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	if len(tolerations) == 0 {
+		glog.V(100).Infof("The tolerations is empty")
```
Comment (achuzhoy) [inline pkg/lso/localvolumeset.go:239]:
```suggestion
		glog.V(100).Infof("The tolerations list is empty")
```

File: pkg/lso/localvolumeset.go [hunk 8]
```diff
@@ -4,36 +4,46 @@ import (
...
 	errorMsg string
 	// api client to interact with the cluster.
-	apiClient *clients.Settings
+	apiClient goclient.Client
 }
 
 // NewLocalVolumeSetBuilder creates new instance of LocalVolumeSetBuilder.
 func NewLocalVolumeSetBuilder(apiClient *clients.Settings, name, nsname string) *LocalVolumeSetBuilder {
-	glog.V(100).Infof("Initializing new %s localVolumeSet structure in %s namespace", name, nsname)
+	glog.V(100).Infof("Initializing new %s localVolumeSet structure in %s namespace",
```
Comment (achuzhoy) [inline pkg/lso/localvolumeset.go:35]:
```suggestion
	glog.V(100).Infof("Initializing new localVolumeSet %s structure in namespace %s",
```

**Review by achuzhoy — DISMISSED** [review]

File: pkg/lso/localvolumeset_test.go [hunk 9]
```diff
@@ -0,0 +1,612 @@
...
+		testLocalVolumeSet *LocalVolumeSetBuilder
+		expectedError      error
+	}{
+		{
+			testLocalVolumeSet: buildValidLocalVolumeSetObjectBuilder(buildLocalVolumeSetClientWithDummyObject()),
+			expectedError:      nil,
+		},
+		{
+			testLocalVolumeSet: buildValidLocalVolumeSetObjectBuilder(clients.GetTestClients(clients.TestClientParams{})),
+			expectedError:      fmt.Errorf("localVolumeSet cannot be deleted because it does not exist"),
```
Comment (kononovn) [inline pkg/lso/localvolumeset_test.go:262]: Please update Delete function.
We don't return error if object doesn't exist. Just print log message and return nil. 
Please refer to:
https://github.com/openshift-kni/eco-goinfra/blob/82c1862d9c98d00fac446c4bf84d37ca5ac36e08/pkg/argocd/applications.go#L162

**Review by kononovn — CHANGES_REQUESTED** [review]
1 comment to fix

**Review by kononovn — APPROVED** [review]
/lgtm

**Pull request merged**